README.md
Fetch
=====

**Fetch это библиотека для чтения электронной почты и вложений, по протоколам POP и IMAP.**

## Установка

Предпочтительный способ установки расширения через [composer](http://getcomposer.org/download/).

Запустить

```sh
php composer.phar require "apexwire/fetch" : "^0.9.0"
```

или добавить

```json
"apexwire/fetch": "^0.9.0"
```

в разделе "require" вашего composer.json

## Применение

Это лишь простой код, чтобы показать, как получить доступ к сообщениям с помощью Fetch. Он использует Fetch
собственный автозагрузка, но он может (и должен быть, если это применимо) заменяется генерируемому
композитором.

```php
    $server = new \Fetch\Server('imap.example.com', 993);
    $server->setAuthentication('dummy', 'dummy');

    $messages = $server->getMessages();
    /** @var $message \Fetch\Message */
    foreach ($messages as $message) {
        echo "Subject: {$message->getSubject()}\nBody: {$message->getMessageBody()}\n";
    }
```

## Лицензия

Этот проект был выпущен под лицензией [BSD-3-Clause](LICENSE).
Подробнее [тут](http://choosealicense.com/licenses/bsd-3-clause).

Copyright © 2016, ApexWire

## Выражение признательности

- Проект является форком [Fetch](https://github.com/tedious/Fetch).


CHANGELOG.md
apexwire/fetch changelog
---------------------------

## v0.10.1 on 2016.07.25

- fix bug: Undefined property: stdClass::$host. Message.php(807)

## v0.10.0 on 2016.05.24

- add function: closeConnection

## v0.9.1 on 2016.05.24

- fix bug: Undefined property: stdClass::$bytes. Attachment.php(102)
- fix bug: Undefined property: stdClass::$subject. Message.php(256)

## v0.9.0 on 2016.05.16

- Add rawEmail in Message
- Add function encodingsIdToString 
- Now rawBody associative array

## v0.8.0 on 2016.04.15

- The minimum version of php 5.4
- Add rawBody in Message
- Add changelog.md
- README.md rename README_EN.md. Add README.md - russian 
- Merge pull request [#147](https://github.com/tedious/Fetch/pull/147) from linniksa/patch-mime-decoder 
- Changes from the pull request [#151](https://github.com/tedious/Fetch/pull/151)

## v0.7.1 on 2 Aug 2015

- Merge pull request [#145](https://github.com/tedious/Fetch/pull/145) from tedious/supress_php_warning
- Suppressed imap_open warning

## v0.6.1 on 8 Jan 2015

- Merge pull request [#109](https://github.com/tedious/Fetch/pull/109) from tedious/release-0.6.1
- Changed function names for cs consistency


## v0.5.3 on 14 Mar 2014

- Merge pull request [#40](https://github.com/tedious/Fetch/pull/40) from bjornpost/fix-multipart-messagebody
- In a multipart email messageBody() keeps headers


## v0.5.2 on 21 Jan 2014

- Merge pull request [#33](https://github.com/tedious/Fetch/pull/33) from tedivm/testing_update
- Separate Mail Server Setup into Development Package


## v0.5.1 on 19 Dec 2013

- Removed autoloader reference


## v0.4.5 on 1 Dec 2013

- Merge pull request [#27](https://github.com/tedious/Fetch/pull/27) from ArabCoders/master
- fixed bug in getting reply_to addresses


## v0.4.4 on 28 Jul 2013

- Merge pull request [#19](https://github.com/tedious/Fetch/pull/19) from codyfletcher/master
- Added support for "bcc" as a type of processed object.


## v0.4.3 on 12 Jul 2013

- Merge pull request [#17](https://github.com/tedious/Fetch/pull/17) from abimus/fix/flags
- Fix behavior on Server::setFlag()


## v0.4.2 on 2 Jul 2013

- Added bug fixed.

## v0.4.1 on 26 Nov 2012

- First release since migrating to github


tests/Fetch/Test/MIMETest.php
<?php

/*
 * This file is part of the Fetch library.
 *
 * (c) Robert Hafner <tedivm@tedivm.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Fetch\Test;

use Fetch\MIME;

/**
 * @package Fetch
 * @author  Robert Hafner <tedivm@tedivm.com>
 * @author  Sergey Linnik <linniksa@gmail.com>
 */
class MIMETest extends \PHPUnit_Framework_TestCase
{
    public function decodeData()
    {
        return [
            [null, null],
            ['Just text', 'Just text'],
            ['Keith Moore <moore@cs.utk.edu>', '=?US-ASCII?Q?Keith_Moore?= <moore@cs.utk.edu>'],
            ['Keld Jørn Simonsen <keld@dkuug.dk>', '=?ISO-8859-1?Q?Keld_J=F8rn_Simonsen?= <keld@dkuug.dk>'],
            ['André Pirard <PIRARD@vm1.ulg.ac.be>', '=?ISO-8859-1?Q?Andr=E9?= Pirard <PIRARD@vm1.ulg.ac.be>'],
            [
                'If you can read this you understand the example.',
                '=?ISO-8859-1?B?SWYgeW91IGNhbiByZWFkIHRoaXMgeW8=?='
                . PHP_EOL .
                '=?ISO-8859-2?B?dSB1bmRlcnN0YW5kIHRoZSBleGFtcGxlLg==?='
            ],
        ];
    }

    /**
     * @dataProvider decodeData
     *
     * @param string $expected
     * @param string $text
     * @param string $charset
     */
    public function testDecode($expected, $text, $charset = 'UTF-8')
    {
        self::assertSame($expected, MIME::decode($text, $charset));
    }
}


tests/Fetch/Test/ServerTest.php
<?php

/*
 * This file is part of the Fetch library.
 *
 * (c) Robert Hafner <tedivm@tedivm.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Fetch\Test;

use Fetch\Server;

/**
 * @package Fetch
 * @author  Robert Hafner <tedivm@tedivm.com>
 */
class ServerTest extends \PHPUnit_Framework_TestCase
{
    public static $num_messages_inbox = 12;

    /**
     * @dataProvider flagsDataProvider
     * @param string $expected server string with %host% placeholder
     * @param integer $port to use (needed to test behavior on port 143 and 993 from constructor)
     * @param array $flags to set/unset ($flag => $value)
     */
    public function testFlags($expected, $port, $flags)
    {
        $server = new Server(TESTING_SERVER_HOST, $port);

        foreach ($flags as $flag => $value) {
            $server->setFlag($flag, $value);
        }

        $this->assertEquals(str_replace('%host%', TESTING_SERVER_HOST, $expected), $server->getServerString());
    }

    public function testFlagOverwrite()
    {
        $server = static::getServer();

        $server->setFlag('TestFlag', 'true');
        $this->assertAttributeContains('TestFlag=true', 'flags', $server);

        $server->setFlag('TestFlag', 'false');
        $this->assertAttributeContains('TestFlag=false', 'flags', $server);
    }

    public function flagsDataProvider()
    {
        return [
            ['{%host%:143/novalidate-cert}', 143, []],
            ['{%host%:143/validate-cert}', 143, ['validate-cert' => true]],
            ['{%host%:143}', 143, ['novalidate-cert' => false]],
            ['{%host%:993/ssl}', 993, []],
            ['{%host%:993}', 993, ['ssl' => false]],
            ['{%host%:100/tls}', 100, ['tls' => true]],
            ['{%host%:100/tls}', 100, ['tls' => true, 'tls' => true]],
            ['{%host%:100/notls}', 100, ['tls' => true, 'notls' => true]],
            ['{%host%:100}', 100, ['ssl' => true, 'ssl' => false]],
            ['{%host%:100/user=foo}', 100, ['user' => 'foo']],
            ['{%host%:100/user=foo}', 100, ['user' => 'foo', 'user' => 'foo']],
            ['{%host%:100/user=bar}', 100, ['user' => 'foo', 'user' => 'bar']],
            ['{%host%:100}', 100, ['user' => 'foo', 'user' => false]],
        ];
    }

    /**
     * @dataProvider connectionDataProvider
     * @param integer $port to use (needed to test behavior on port 143 and 993 from constructor)
     * @param array $flags to set/unset ($flag => $value)
     * @param string $message Assertion message
     */
    public function testConnection($port, $flags, $message)
    {
        $server = new Server(TESTING_SERVER_HOST, $port);
        $server->setAuthentication(TEST_USER, TEST_PASSWORD);

        foreach ($flags as $flag => $value) {
            $server->setFlag($flag, $value);
        }

        $imapSteam = $server->getImapStream();
        $this->assertInternalType('resource', $imapSteam, $message);
    }

    public function connectionDataProvider()
    {
        return [
            [143, [], 'Connects with default settings.'],
            [993, ['novalidate-cert' => true], 'Connects over SSL (self signed).'],
        ];
    }

    public function testNumMessages()
    {
        $server = static::getServer();
        $numMessages = $server->numMessages();
        $this->assertEquals(self::$num_messages_inbox, $numMessages);
        $this->assertEquals(0, $server->numMessages('DOESNOTEXIST' . time()));
    }

    public function testGetMessages()
    {
        $server = static::getServer();
        $messages = $server->getMessages(5);

        $this->assertCount(5, $messages, 'Five messages returned');
        foreach ($messages as $message) {
            $this->assertInstanceOf('\Fetch\Message', $message, 'Returned values are Messages');
        }
    }

    public function testGetMessagesOrderedByDateAsc()
    {
        $server = static::getServer();
        $messages = $server->getOrderedMessages(SORTDATE, false, 2);

        $this->assertCount(2, $messages, 'Two messages returned');
        $this->assertGreaterThan($messages[0]->getDate(), $messages[1]->getDate(), 'Messages in ascending order');
    }

    public function testGetMessagesOrderedByDateDesc()
    {
        $server = static::getServer();
        $messages = $server->getOrderedMessages(SORTDATE, true, 2);

        $this->assertCount(2, $messages, 'Two messages returned');
        $this->assertLessThan($messages[0]->getDate(), $messages[1]->getDate(), 'Messages in descending order');
    }

    public function testGetMailBox()
    {
        $server = static::getServer();
        $this->assertEquals('', $server->getMailBox());
        $this->assertTrue($server->setMailBox('Sent'));
        $this->assertEquals('Sent', $server->getMailBox());
    }

    public function testSetMailBox()
    {
        $server = static::getServer();

        $this->assertTrue($server->setMailBox('Sent'));
        $this->assertEquals('Sent', $server->getMailBox());

        $this->assertTrue($server->setMailBox('Flagged Email'));
        $this->assertEquals('Flagged Email', $server->getMailBox());

        $this->assertFalse($server->setMailBox('Cheese'));

        $this->assertTrue($server->setMailBox(''));
        $this->assertEquals('', $server->getMailBox());
    }

    public function testHasMailBox()
    {
        $server = static::getServer();

        $this->assertTrue($server->hasMailBox('Sent'), 'Has mailbox "Sent"');
        $this->assertTrue($server->hasMailBox('Flagged Email'), 'Has mailbox "Flagged Email"');
        $this->assertFalse($server->hasMailBox('Cheese'), 'Does not have mailbox "Cheese"');
    }

    public function testListMailBoxes()
    {
        $server = static::getServer();
        $spec = sprintf('{%s:143/novalidate-cert}', TESTING_SERVER_HOST);

        $list = $server->listMailboxes('*');
        $this->assertContains($spec . 'Sent', $list, 'Has mailbox "Sent"');
        $this->assertNotContains($spec . 'Cheese', $list, 'Does not have mailbox "Cheese"');
    }

    public function testCreateMailbox()
    {
        $server = static::getServer();

        $this->assertFalse($server->hasMailBox('Cheese'), 'Does not have mailbox "Cheese"');
        $this->assertTrue($server->createMailBox('Cheese'), 'createMailbox returns true.');
        $this->assertTrue($server->hasMailBox('Cheese'), 'Mailbox "Cheese" was created');
    }

    public function testDeleteMailbox()
    {
        $server = static::getServer();
        $this->assertTrue($server->hasMailBox('Cheese'), 'Does have mailbox "Cheese"');
        $this->assertTrue($server->deleteMailBox('Cheese'), 'deleteMailBox returns true.');
        $this->assertFalse($server->hasMailBox('Cheese'), 'Mailbox "Cheese" was deleted');
    }

    /**
     * @expectedException \RuntimeException
     */
    public function testSetOptionsException()
    {
        $server = static::getServer();
        $server->setOptions('purple');
    }

    public function testSetOptions()
    {
        $server = static::getServer();
        $server->setOptions(5);
        $this->assertAttributeEquals(5, 'options', $server);
    }

    public function testExpunge()
    {
        $server = static::getServer();
        $message = $server->getMessageByUid(12);

        $this->assertInstanceOf('\Fetch\Message', $message, 'Message exists');

        $message->delete();

        $this->assertInstanceOf('\Fetch\Message', $server->getMessageByUid(12),
            'Message still present after being deleted but before being expunged.');

        $server->expunge();

        $this->assertFalse($server->getMessageByUid(12), 'Message successfully expunged');
    }

    public static function getServer()
    {
        $server = new Server(TESTING_SERVER_HOST, 143);
        $server->setAuthentication(TEST_USER, TEST_PASSWORD);

        return $server;
    }
}


src/Fetch/Attachment.php
<?php

/*
 * This file is part of the Fetch package.
 *
 * (c) Robert Hafner <tedivm@tedivm.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Fetch;

/**
 * This library is a wrapper around the Imap library functions included in php. This class wraps around an attachment
 * in a message, allowing developers to easily save or display attachments.
 *
 * @package Fetch
 * @author  Robert Hafner <tedivm@tedivm.com>
 */
class Attachment
{

    /**
     * This is the structure object for the piece of the message body that the attachment is located it.
     *
     * @var \stdClass
     */
    protected $structure;

    /**
     * This is the unique identifier for the message this attachment belongs to.
     *
     * @var int
     */
    protected $messageId;

    /**
     * This is the ImapResource.
     *
     * @var resource
     */
    protected $imapStream;

    /**
     * This is the id pointing to the section of the message body that contains the attachment.
     *
     * @var int
     */
    protected $partId;

    /**
     * This is the attachments filename.
     *
     * @var string
     */
    protected $filename;

    /**
     * This is the size of the attachment.
     *
     * @var int
     */
    protected $size;

    /**
     * This stores the data of the attachment so it doesn't have to be retrieved from the server multiple times. It is
     * only populated if the getData() function is called and should not be directly used.
     *
     * @internal
     * @var array
     */
    protected $data;

    /**
     * This function takes in an ImapMessage, the structure object for the particular piece of the message body that the
     * attachment is located at, and the identifier for that body part. As a general rule you should not be creating
     * instances of this yourself, but rather should get them from an ImapMessage class.
     *
     * @param Message $message
     * @param \stdClass $structure
     * @param string $partIdentifier
     */
    public function __construct(Message $message, $structure, $partIdentifier = null)
    {
        $this->messageId = $message->getUid();
        $this->imapStream = $message->getImapBox()->getImapStream();
        $this->structure = $structure;

        if (isset($partIdentifier)) {
            $this->partId = $partIdentifier;
        }

        $parameters = Message::getParametersFromStructure($structure);

        if (isset($parameters['filename'])) {
            $this->setFileName($parameters['filename']);
        } elseif (isset($parameters['name'])) {
            $this->setFileName($parameters['name']);
        }

        $this->size = empty($structure->bytes) ? -1 : $structure->bytes;

        $this->mimeType = Message::typeIdToString($structure->type);

        if (isset($structure->subtype)) {
            $this->mimeType .= '/' . strtolower($structure->subtype);
        }

        $this->encoding = $structure->encoding;
    }

    /**
     * This function returns the data of the attachment. Combined with getMimeType() it can be used to directly output
     * data to a browser.
     *
     * @return string
     */
    public function getData()
    {
        if (!isset($this->data)) {
            $messageBody = isset($this->partId) ?
                imap_fetchbody($this->imapStream, $this->messageId, $this->partId, FT_UID)
                : imap_body($this->imapStream, $this->messageId, FT_UID);

            $messageBody = Message::decode($messageBody, $this->encoding);
            $this->data = $messageBody;
        }

        return $this->data;
    }

    /**
     * This returns the filename of the attachment, or false if one isn't given.
     *
     * @return string
     */
    public function getFileName()
    {
        return (isset($this->filename)) ? $this->filename : false;
    }

    /**
     * This function returns the mimetype of the attachment.
     *
     * @return string
     */
    public function getMimeType()
    {
        return $this->mimeType;
    }

    /**
     * This returns the size of the attachment.
     *
     * @return int
     */
    public function getSize()
    {
        return $this->size;
    }

    /**
     * This function returns the object that contains the structure of this attachment.
     *
     * @return \stdClass
     */
    public function getStructure()
    {
        return $this->structure;
    }

    /**
     * This function saves the attachment to the passed directory, keeping the original name of the file.
     *
     * @param  string $path
     * @return bool
     */
    public function saveToDirectory($path)
    {
        $path = rtrim($path, '/') . '/';

        if (is_dir($path)) {
            return $this->saveAs($path . $this->getFileName());
        }

        return false;
    }

    /**
     * This function saves the attachment to the exact specified location.
     *
     * @param  string $path
     * @return bool
     */
    public function saveAs($path)
    {
        $dirName = dirname($path);
        if (file_exists($path)) {
            if (!is_writable($path)) {
                return false;
            }
        } elseif (!is_dir($dirName) || !is_writable($dirName)) {
            return false;
        }

        if (($filePointer = fopen($path, 'w')) == false) {
            return false;
        }

        switch ($this->encoding) {
            case 3: //base64
                $streamFilter = stream_filter_append($filePointer, 'convert.base64-decode', STREAM_FILTER_WRITE);
                break;

            case 4: //quoted-printable
                $streamFilter = stream_filter_append($filePointer, 'convert.quoted-printable-decode',
                    STREAM_FILTER_WRITE);
                break;

            default:
                $streamFilter = null;
        }

        // Fix an issue causing server to throw an error
        // See: https://github.com/tedious/Fetch/issues/74 for more details
        imap_fetchbody($this->imapStream, $this->messageId, $this->partId ?: 1, FT_UID);
        $result = imap_savebody($this->imapStream, $filePointer, $this->messageId, $this->partId ?: 1, FT_UID);

        if ($streamFilter) {
            stream_filter_remove($streamFilter);
        }

        fclose($filePointer);

        return $result;
    }

    /**
     * @param $text
     */
    protected function setFileName($text)
    {
        $this->filename = MIME::decode($text, Message::$charset);
    }
}


src/Fetch/Server.php
<?php

/*
 * This file is part of the Fetch package.
 *
 * (c) Robert Hafner <tedivm@tedivm.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Fetch;

/**
 * This library is a wrapper around the Imap library functions included in php. This class in particular manages a
 * connection to the server (imap, pop, etc) and allows for the easy retrieval of stored messages.
 *
 * @package Fetch
 * @author  Robert Hafner <tedivm@tedivm.com>
 */
class Server
{
    /**
     * When SSL isn't compiled into PHP we need to make some adjustments to prevent soul crushing annoyances.
     *
     * @var bool
     */
    public static $sslEnable = true;

    /**
     * These are the flags that depend on ssl support being compiled into imap.
     *
     * @var array
     */
    public static $sslFlags = ['ssl', 'validate-cert', 'novalidate-cert', 'tls', 'notls'];

    /**
     * This is used to prevent the class from putting up conflicting tags. Both directions- key to value, value to key-
     * are checked, so if "novalidate-cert" is passed then "validate-cert" is removed, and vice-versa.
     *
     * @var array
     */
    public static $exclusiveFlags = ['validate-cert' => 'novalidate-cert', 'tls' => 'notls'];

    /**
     * This is the domain or server path the class is connecting to.
     *
     * @var string
     */
    protected $serverPath;

    /**
     * This is the name of the current mailbox the connection is using.
     *
     * @var string
     */
    protected $mailbox = '';

    /**
     * This is the username used to connect to the server.
     *
     * @var string
     */
    protected $username;

    /**
     * This is the password used to connect to the server.
     *
     * @var string
     */
    protected $password;

    /**
     * This is an array of flags that modify how the class connects to the server. Examples include "ssl" to enforce a
     * secure connection or "novalidate-cert" to allow for self-signed certificates.
     *
     * @link http://us.php.net/manual/en/function.imap-open.php
     * @var array
     */
    protected $flags = [];

    /**
     * This is the port used to connect to the server
     *
     * @var int
     */
    protected $port;

    /**
     * This is the set of options, represented by a bitmask, to be passed to the server during connection.
     *
     * @var int
     */
    protected $options = 0;

    /**
     * This is the set of connection parameters
     *
     * @var array
     */
    protected $params = [];

    /**
     * This is the resource connection to the server. It is required by a number of imap based functions to specify how
     * to connect.
     *
     * @var resource
     */
    protected $imapStream;

    /**
     * This is the name of the service currently being used. Imap is the default, although pop3 and nntp are also
     * options
     *
     * @var string
     */
    protected $service = 'imap';

    /**
     * This constructor takes the location and service thats trying to be connected to as its arguments.
     *
     * @param string $serverPath
     * @param null|int $port
     * @param null|string $service
     */
    public function __construct($serverPath, $port = 143, $service = 'imap')
    {
        $this->serverPath = $serverPath;

        $this->port = $port;

        switch ($port) {
            case 143:
                $this->setFlag('novalidate-cert');
                break;

            case 993:
                $this->setFlag('ssl');
                break;
        }

        $this->service = $service;
    }

    /**
     * This function sets the username and password used to connect to the server.
     *
     * @param string $username
     * @param string $password
     * @param bool $tryFasterAuth tries to auth faster by disabling GSSAPI & NTLM auth methods (set to false if you use either of these auth methods)
     */
    public function setAuthentication($username, $password, $tryFasterAuth = true)
    {
        $this->username = $username;
        $this->password = $password;
        if ($tryFasterAuth) {
            $this->setParam('DISABLE_AUTHENTICATOR', ['GSSAPI', 'NTLM']);
        }
    }

    /**
     * This function sets the mailbox to connect to.
     *
     * @param  string $mailbox
     * @return bool
     */
    public function setMailBox($mailbox = '')
    {
        if (!$this->hasMailBox($mailbox)) {
            return false;
        }

        $this->mailbox = $mailbox;
        if (isset($this->imapStream)) {
            $this->setImapStream();
        }

        return true;
    }

    /**
     * @return string
     */
    public function getMailBox()
    {
        return $this->mailbox;
    }

    /**
     * This function sets or removes flag specifying connection behavior. In many cases the flag is just a one word
     * deal, so the value attribute is not required. However, if the value parameter is passed false it will clear that
     * flag.
     *
     * @param string $flag
     * @param null|string|bool $value
     */
    public function setFlag($flag, $value = null)
    {
        if (!self::$sslEnable && in_array($flag, self::$sslFlags)) {
            return;
        }

        if (isset(self::$exclusiveFlags[$flag])) {
            $kill = self::$exclusiveFlags[$flag];
        } elseif ($index = array_search($flag, self::$exclusiveFlags)) {
            $kill = $index;
        }

        if (isset($kill) && false !== $index = array_search($kill, $this->flags)) {
            unset($this->flags[$index]);
        }

        $index = array_search($flag, $this->flags);
        if (isset($value) && $value !== true) {
            if ($value == false && $index !== false) {
                unset($this->flags[$index]);
            } elseif ($value != false) {
                $match = preg_grep('/' . $flag . '/', $this->flags);
                if (reset($match)) {
                    $this->flags[key($match)] = $flag . '=' . $value;
                } else {
                    $this->flags[] = $flag . '=' . $value;
                }
            }
        } elseif ($index === false) {
            $this->flags[] = $flag;
        }
    }

    /**
     * This function is used to set various options for connecting to the server.
     *
     * @param  int $bitmask
     * @throws \Exception
     */
    public function setOptions($bitmask = 0)
    {
        if (!is_numeric($bitmask)) {
            throw new \RuntimeException('Function requires numeric argument.');
        }

        $this->options = $bitmask;
    }

    /**
     * This function is used to set connection parameters
     *
     * @param string $key
     * @param string $value
     */
    public function setParam($key, $value)
    {
        $this->params[$key] = $value;
    }

    /**
     * This function gets the current saved imap resource and returns it.
     *
     * @return resource
     */
    public function getImapStream()
    {
        if (empty($this->imapStream)) {
            $this->setImapStream();
        }

        return $this->imapStream;
    }

    /**
     * This function takes in all of the connection date (server, port, service, flags, mailbox) and creates the string
     * thats passed to the imap_open function.
     *
     * @return string
     */
    public function getServerString()
    {
        $mailboxPath = $this->getServerSpecification();

        if (isset($this->mailbox)) {
            $mailboxPath .= $this->mailbox;
        }

        return $mailboxPath;
    }

    /**
     * Returns the server specification, without adding any mailbox.
     *
     * @return string
     */
    protected function getServerSpecification()
    {
        $mailboxPath = '{' . $this->serverPath;

        if (isset($this->port)) {
            $mailboxPath .= ':' . $this->port;
        }

        if ($this->service != 'imap') {
            $mailboxPath .= '/' . $this->service;
        }

        foreach ($this->flags as $flag) {
            $mailboxPath .= '/' . $flag;
        }

        $mailboxPath .= '}';

        return $mailboxPath;
    }

    /**
     * This function creates or reopens an imapStream when called.
     */
    protected function setImapStream()
    {
        if (!empty($this->imapStream)) {
            if (!imap_reopen($this->imapStream, $this->getServerString(), $this->options, 1)) {
                throw new \RuntimeException(imap_last_error());
            }
        } else {
            $imapStream = @imap_open($this->getServerString(), $this->username, $this->password, $this->options, 1,
                $this->params);

            if ($imapStream === false) {
                throw new \RuntimeException(imap_last_error());
            }

            $this->imapStream = $imapStream;
        }
    }

    /**
     * This function close an imapStream when called.
     */
    public function closeConnection()
    {
        if (!empty($this->imapStream)) {
            if (!imap_close($this->imapStream)) {
                throw new \RuntimeException(imap_last_error());
            }
        }
    }

    /**
     * This returns the number of messages that the current mailbox contains.
     *
     * @param  string $mailbox
     * @return int
     */
    public function numMessages($mailbox = '')
    {
        $cnt = 0;
        if ($mailbox === '') {
            $cnt = imap_num_msg($this->getImapStream());
        } elseif ($this->hasMailbox($mailbox) && $mailbox !== '') {
            $oldMailbox = $this->getMailBox();
            $this->setMailbox($mailbox);
            $cnt = $this->numMessages();
            $this->setMailbox($oldMailbox);
        }

        return ((int)$cnt);
    }

    /**
     * This function returns an array of ImapMessage object for emails that fit the criteria passed. The criteria string
     * should be formatted according to the imap search standard, which can be found on the php "imap_search" page or in
     * section 6.4.4 of RFC 2060
     *
     * @link http://us.php.net/imap_search
     * @link http://www.faqs.org/rfcs/rfc2060
     * @param  string $criteria
     * @param  null|int $limit
     * @return array    An array of ImapMessage objects
     */
    public function search($criteria = 'ALL', $limit = null)
    {
        if ($results = imap_search($this->getImapStream(), $criteria, SE_UID)) {
            if (isset($limit) && count($results) > $limit) {
                $results = array_slice($results, 0, $limit);
            }

            $messages = [];

            foreach ($results as $messageId) {
                $messages[] = new Message($messageId, $this);
            }

            return $messages;
        } else {
            return [];
        }
    }

    /**
     * This function returns the recently received emails as an array of ImapMessage objects.
     *
     * @param  null|int $limit
     * @return array    An array of ImapMessage objects for emails that were recently received by the server.
     */
    public function getRecentMessages($limit = null)
    {
        return $this->search('Recent', $limit);
    }

    /**
     * Returns the emails in the current mailbox as an array of ImapMessage objects.
     *
     * @param  null|int $limit
     * @return Message[]
     */
    public function getMessages($limit = null)
    {
        $numMessages = $this->numMessages();

        if (isset($limit) && is_numeric($limit) && $limit < $numMessages) {
            $numMessages = $limit;
        }

        if ($numMessages < 1) {
            return [];
        }

        $stream = $this->getImapStream();
        $messages = [];
        for ($i = 1; $i <= $numMessages; $i++) {
            $uid = imap_uid($stream, $i);
            $messages[] = new Message($uid, $this);
        }

        return $messages;
    }

    /**
     * Returns the emails in the current mailbox as an array of ImapMessage objects
     * ordered by some ordering
     *
     * @see    http://php.net/manual/en/function.imap-sort.php
     * @param  int $orderBy
     * @param  bool $reverse
     * @param  int $limit
     * @return Message[]
     */
    public function getOrderedMessages($orderBy, $reverse, $limit)
    {
        $msgIds = imap_sort($this->getImapStream(), $orderBy, $reverse ? 1 : 0, SE_UID);

        return array_map([$this, 'getMessageByUid'], array_slice($msgIds, 0, $limit));
    }

    /**
     * Returns the requested email or false if it is not found.
     *
     * @param  int $uid
     * @return Message|bool
     */
    public function getMessageByUid($uid)
    {
        try {
            $message = new Message($uid, $this);

            return $message;
        } catch (\Exception $e) {
            return false;
        }
    }

    /**
     * This function removes all of the messages flagged for deletion from the mailbox.
     *
     * @return bool
     */
    public function expunge()
    {
        return imap_expunge($this->getImapStream());
    }

    /**
     * Checks if the given mailbox exists.
     *
     * @param $mailbox
     *
     * @return bool
     */
    public function hasMailBox($mailbox)
    {
        return (boolean)$this->getMailBoxDetails($mailbox);
    }

    /**
     * Return information about the mailbox or mailboxes
     *
     * @param $mailbox
     *
     * @return array
     */
    public function getMailBoxDetails($mailbox)
    {
        return imap_getmailboxes(
            $this->getImapStream(),
            $this->getServerString(),
            $this->getServerSpecification() . $mailbox
        );
    }

    /**
     * Creates the given mailbox.
     *
     * @param $mailbox
     *
     * @return bool
     */
    public function createMailBox($mailbox)
    {
        return imap_createmailbox($this->getImapStream(), $this->getServerSpecification() . $mailbox);
    }

    /**
     * List available mailboxes
     *
     * @param string $pattern
     *
     * @return array
     */
    public function listMailBoxes($pattern = '*')
    {
        return imap_list($this->getImapStream(), $this->getServerSpecification(), $pattern);
    }

    /**
     * Deletes the given mailbox.
     *
     * @param $mailbox
     *
     * @return bool
     */
    public function deleteMailBox($mailbox)
    {
        return imap_deletemailbox($this->getImapStream(), $this->getServerSpecification() . $mailbox);
    }
}
